lib/rules/no-shadow.js

```diff
@@ -54,9 +69,95 @@ module.exports = {
+            return types.includes(node.type);
+        }
+
+        /**
+         * Searches from the current node through its ancestry to find a matching node.
+         * @param {ASTNode} node a node to get.
+         * @param {MatchCallback} match a callback that checks whether or not the node verifies its condition or not.
+         * @param {StopSearchingCallback} stopSearching a callback that checks whether or not the searching should be stopped or not.
+         * @returns {ASTNode} the matching node.
+         */
+        function findSelfOrAncestor(node, match, stopSearching) {
+            let currentNode = node;
+
+            while (currentNode && !match(currentNode)) {
+                if (stopSearching && stopSearching(currentNode)) {
+                    return null;
+                }
+                currentNode = currentNode.parent;
+            }
+            return currentNode;
+        }
+
+        /**
+         * Finds function's outer scope.
+         * @param {Scope} scope Function's own scope.
+         * @returns {Scope} Function's outer scope.
+         */
+        function getOuterScope(scope) {
+            const upper = scope.upper;
+
+            if (upper.type === "function-expression-name") {
+                return upper.upper;
+            }
+            return upper;
+        }
+
+        /**
+         * Checks if a variable and a shadowedVariable have the same init pattern ancestor.
+         * @param {Object} variable a variable to check.
+         * @param {Object} shadowedVariable a shadowedVariable to check.
+         * @returns {boolean} Whether or not the variable and the shadowedVariable have the same init pattern ancestor.
+         */
+        function isInitPatternNode(variable, shadowedVariable) {
+            const outerDef = shadowedVariable.defs[0];
+            const { variableScope } = variable.scope;
+
+            if (!(FUN_NODES_TYPES.includes(variableScope.block.type) && getOuterScope(variableScope) === shadowedVariable.scope)) {
+                return false;
+            }
+
+            const fun = variableScope.block;
+            const { parent } = fun;
+
+            const callExpression = findSelfOrAncestor(
+                parent,
+                node => isNodeMatchesTypes(node, CALL_EXP_NODE_TYPE),
+                node => isNodeMatchesTypes(node, FUN_NODES_TYPES)
+            );
+
+            if (!callExpression) {
+                return false;
+            }
+
+            const shadowingVariableAncestor = findSelfOrAncestor(
+                callExpression.parent,
+                node => isNodeMatchesTypes(node, INIT_NODES_TYPES)
+            );
+
+            const shadowingVariable = outerDef && outerDef.name;
+
+            const shadowedVariableAncestor = findSelfOrAncestor(
+                shadowingVariable,
+                node => node === shadowingVariableAncestor
+            );
+
+            return !!shadowedVariableAncestor;
```

Comment:
This could find a common ancestor that isn't the one that declares the shadowed variable.

```js
/* eslint no-shadow: [2, { ignoreOnInitialization: true }] */

const foo = () => {
    let x;
    bar(x => x); // not reported, but it should be?
}

```

Maybe we could borrow the code that checks if a node is in variable's initializer from the `no-use-before-define` rule:

https://github.com/eslint/eslint/blob/355b23d0c4e050be4e53292f552a47c10ec6e00e/lib/rules/no-use-before-define.js#L190-L212